src/claimtrie.cpp

```diff
@@ -963,164 +624,41 @@ bool CClaimTrie::updateTakeoverHeight(const std::string& name, int nTakeoverHeig
+bool CClaimTrie::WriteToDisk()
 {
-    for (expirationQueueType::iterator itQueue = dirtySupportExpirationQueueRows.begin(); itQueue != dirtySupportExpirationQueueRows.end(); ++itQueue)
-    {
-        if (itQueue->second.empty())
-        {
-            batch.Erase(std::make_pair(SUPPORT_EXP_QUEUE_ROW, itQueue->first));
-        }
-        else
-        {
-            batch.Write(std::make_pair(SUPPORT_EXP_QUEUE_ROW, itQueue->first), itQueue->second);
+    for (nodeCacheType::iterator itcache = dirtyNodes.begin(); itcache != dirtyNodes.end(); ++itcache) {
+        CClaimTrieNode* pNode = itcache->second;
+        uint32_t num_claims = pNode ? pNode->claims.size() : 0;
+        LogPrintf("%s: Writing %s to disk with %d claims\n", __func__, itcache->first, num_claims);
+        if (pNode) {
+            CClaimTrieNode copy(*pNode);
+            db.updateQueueRow(itcache->first, copy);
+        } else {
+            CClaimTrieNode emptyNode;
+            db.updateQueueRow(itcache->first, emptyNode);
         }
     }
-}
 
-bool CClaimTrie::WriteToDisk()
-{
-    CDBBatch batch(&db.GetObfuscateKey());
-    for (nodeCacheType::iterator itcache = dirtyNodes.begin(); itcache != dirtyNodes.end(); ++itcache)
-        BatchWriteNode(batch, itcache->first, itcache->second);
     dirtyNodes.clear();
-    BatchWriteQueueRows(batch);
-    dirtyQueueRows.clear();
-    BatchWriteQueueNameRows(batch);
-    dirtyQueueNameRows.clear();
-    BatchWriteExpirationQueueRows(batch);
-    dirtyExpirationQueueRows.clear();
-    BatchWriteSupportNodes(batch);
-    dirtySupportNodes.clear();
-    BatchWriteSupportQueueRows(batch);
-    dirtySupportQueueRows.clear();
-    BatchWriteSupportQueueNameRows(batch);
-    dirtySupportQueueNameRows.clear();
-    BatchWriteSupportExpirationQueueRows(batch);
-    dirtySupportExpirationQueueRows.clear();
-    batch.Write(HASH_BLOCK, hashBlock);
-    batch.Write(CURRENT_HEIGHT, nCurrentHeight);
-    return db.WriteBatch(batch);
+    db.writeQueues();
+    db.Write(HASH_BLOCK, hashBlock);
+    db.Write(CURRENT_HEIGHT, nCurrentHeight);
+    return db.Sync();
 }
 
 bool CClaimTrie::InsertFromDisk(const std::string& name, CClaimTrieNode* node)
 {
-    if (name.size() == 0)
-    {
+    if (name.size() == 0) {
         root = *node;
         return true;
     }
     CClaimTrieNode* current = &root;
-    for (std::string::const_iterator itname = name.begin(); itname + 1 != name.end(); ++itname)
-    {
+    for (std::string::const_iterator itname = name.begin(); itname + 1 != name.end(); ++itname) {
         nodeMapType::iterator itchild = current->children.find(*itname);
-        if (itchild == current->children.end())
-            return false;
+        if (itchild == current->children.end()) return false;
         current = itchild->second;
     }
-    current->children[name[name.size()-1]] = node;
+    current->children[name[name.size() - 1]] = new CClaimTrieNode(*node);
```

Comment:
Why do we need to clone this? That looks like a good way to make a memory leak.

Reply:
It should, i'll provide memleak free RR next to this.